engine/scene/sceneloader.cpp

```diff
@@ -0,0 +1,198 @@
+    };
+
+    mComponentMappings["ScoreController"] = [](GameObject* gameObject) {
+        return gameObject->AddComponent<ScoreController>();
+    };
+
+    mComponentMappings["TitleScreenController"] = [](GameObject* gameObject) {
+        return gameObject->AddComponent<TitleScreenController>();
+    };
+
+    mComponentMappings["SettingsScreenController"] = [](GameObject* gameObject) {
+        return gameObject->AddComponent<SettingsScreenController>();
+    };
+}
+
+void SceneLoader::PreLoadScenes()
+{
+    std::vector<std::string> sceneNames = Config::GetValue("scenes", std::vector<std::string> {});
+
+    for (const auto& sceneName : sceneNames)
+    {
+        std::ifstream sceneFile { sceneName };
+
+        if (!sceneFile.is_open())
+        {
+            LogError("Could not open scene file: {}", sceneName);
+            ASSERT(false);
+            continue;
+        }
+
+        nlohmann::json sceneJson;
+        sceneFile >> sceneJson;
+
+        if (sceneJson.is_null())
+        {
+            LogError("Scene file is empty: {}", sceneName);
+            ASSERT(false);
+            continue;
+        }
+
+        // Validating the scene file here saves some error checking later
+        ASSERT(sceneJson.is_object());
+        ASSERT(sceneJson.contains("name") && sceneJson["name"].is_string());
+        const std::string sceneName = sceneJson["name"];
+
+        ASSERT(sceneJson.contains("objects") && sceneJson["objects"].is_array());
+        for (const auto& gameObjectJson : sceneJson["objects"])
+        {
+            ASSERT(gameObjectJson.is_object());
+            ASSERT(gameObjectJson.contains("name") && gameObjectJson["name"].is_string());
+            ASSERT(gameObjectJson.contains("components") && gameObjectJson["components"].is_array());
+            for (const auto& componentJson : gameObjectJson["components"])
+            {
+                ASSERT(componentJson.is_object());
+                ASSERT(componentJson.contains("type") && componentJson["type"].is_string());
+                if (mComponentMappings.find(componentJson["type"]) == mComponentMappings.end())
+                {
+                    LogError("Unknown component type: {}", componentJson["type"]);
+                    ASSERT(false);
+                }
+            }
+        }
+
+        ASSERT(mScenes.find(sceneName) == mScenes.end());
+        mScenes[sceneName] = sceneJson;
+    }
+}
+
+std::vector<std::unique_ptr<GameObject>> SceneLoader::LoadScene(const std::string& sceneName)
+{
+    if (mScenes.find(sceneName) == mScenes.end())
+    {
+        LogError("Scene {} does not exist!", sceneName);
+        ASSERT(false);
+        return {};
+    }
+
+    const nlohmann::json& sceneJson = mScenes[sceneName];
+    return LoadSceneFromJson(sceneJson);
+}
+
+std::vector<std::unique_ptr<GameObject>> SceneLoader::LoadSceneFromJson(const nlohmann::json& sceneJson)
+{
+    std::vector<std::unique_ptr<GameObject>> gameObjects {};
```

Comment:
You use this type enough that it may be worth abstracting to a well named `using`